flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,63 +525,105 @@ class ExprLowering {
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 16) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else {
+        // convert everything else to double
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
       }
-      llvm_unreachable("real constant has no value");
     } else if constexpr (TC == Fortran::lower::ComplexCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value()) {
-        using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
-        return genval(Fortran::evaluate::ComplexConstructor<KIND>{
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->REAL()}},
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->AIMAG()}}});
-      }
-      llvm_unreachable("array of complex unhandled");
-    } else if constexpr (TC == Fortran::lower::CharacterCat) {
-      return genCharLit<KIND>(con.GetScalarValue().value(), con.LEN());
+      using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
+      return genval(Fortran::evaluate::ComplexConstructor<KIND>{
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.REAL()}},
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.AIMAG()}}});
     } else {
       llvm_unreachable("unhandled constant");
     }
   }
 
+  template <Fortran::common::TypeCategory TC, int KIND>
+  mlir::Value genArrayLit(
+      const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
+          &con) {
+    // TODO:
+    // - Multi-dimensional array
+    using T = Fortran::evaluate::Type<TC, KIND>;
+    if (con.Rank() > 1) {
+      TODO(); // "Multi dimensional array not yet supported"
+    }
+    std::vector<Fortran::evaluate::Scalar<T>> values;
+    auto idxTy = builder.getIndexType();
+    fir::InsertValueOp res;
+    std::size_t counter{0};
+    if constexpr ((TC == Fortran::lower::IntegerCat) ||
+                  (TC == Fortran::lower::RealCat) ||
+                  (TC == Fortran::lower::ComplexCat) ||
+                  (TC == Fortran::lower::LogicalCat)) {
+      values = con.GetArrayValue();
+    } else {
+      llvm_unreachable("unhandled constant");
+    }
+    fir::SequenceType::Shape shape{(int64_t)(values.size())};
+    mlir::Type arrayTy =
+        fir::SequenceType::get(shape, converter.genType(TC, KIND));
+    auto array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
+    for (const Fortran::evaluate::Scalar<T> &value : values) {
+      auto idx = builder.createIntegerConstant(idxTy, counter);
+      auto constant = genScalar<TC, KIND>(value);
+      res = builder.create<fir::InsertValueOp>(
```

Comment:
You can just re-use array here instead of res to avoid the ternary (you would need to declare `mlir::Value array` instead of `auto` above though).

Reply:
This was my first approach as well. However, I get this error
`error: ‘class mlir::Value’ has no member named ‘getResult’; did you mean ‘getUseList’?`

Reply:
Right, you do not need to write `array.getResult()` in the `builder.create<fir::InsertValueOp>` if `array` is already declared as an `mlir::Value`. `getResult()/getResult(x)` is only used on Operations (the things returned by `builder.create<xxOp>`) to get the result of the operation (which is an `mlir::Value`). But most of the time you do not need it because the default cast from an Operation to a mlir::Value is equivalent to calling `getResult()`.

So:
```
auto array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
// array has fir::UndefOp type, you can call getResult() on it to get the resulting
// mlir value
array.getResult() // OK, returns mlir::Value

auto array1 = builder.create<fir::UndefOp>(getLoc(), arrayTy).getResult();
// array1 has mlir::Value type, it does not make sens anymore to call getResult() on it.

mlir::Value array2 = builder.create<fir::UndefOp>(getLoc(), arrayTy);
// This case is equivalent to the previous one, for one result operation, the `getResult()`
// is called while casting the operation to mlir::Value.
```

In function calls that expect an `mlir::Value`, the cast from one result operation to the returned value is also automatic:

```
void foo(mlir::Value);
auto array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
foo(array); // OK, array is casted to its result mlir::Value.
```

So their is usually little reason to explicitly call getResult, unless the operation is returning several values or the context is ambiguous/c++ cannot insert the cast on its own.

All that to say, you should be able to write:
```
mlir::Value array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
for (const Fortran::evaluate::Scalar<T> &value : values) {
    auto idx = builder.createIntegerConstant(idxTy, counter);
    auto constant = genScalar<TC, KIND>(value);
    array = builder.create<fir::InsertValueOp>(getLoc(), arrayTy, array, constant, idx);
  }
  return array;
```

Reply:
Thanks for the wonderful explanation. Your suggestion worked!